In [ ]:
print("hello world")

In [ ]:
import pandas as pd
import numpy as np

text cell

![SPARK Banner](https://i.imgur.com/3vCmTns.png)

# SPARK | Day 1 | July 12th, 2021

The agenda for today:

1. About the Tools

2. Introduction to Pandas (What is it?, Reading Excel and CSV Files, Viewing DataFrames)
3. Indexing DataFrames
4. Cleaning DataFrames
5. Version Control - Introduction to Git
6. How to Give Feedback




# About the tools





## Google Colab
<img src="https://thumbor.forbes.com/thumbor/960x0/https%3A%2F%2Fblogs-images.forbes.com%2Fgregoryferenstein%2Ffiles%2F2019%2F08%2Fscreenshot-1200x688.jpg" alt="Colab" width="500"/>

**Google Colab** is an online version of Jupyter Notebooks. It has cells where you can write notes and execute lines of code. We will be running the programming language called **Python** within these notebooks.

To use Google Colab, you will need to have [Google Chrome](https://www.google.com/chrome/) intalled on your computer. 

You will also need a Google Account to be able to access [Google Drive](https://drive.google.com), where the Colab notebooks can be accessed. 

# Pandas









## What is Pandas?

<img src="https://miro.medium.com/max/2550/1*6d5dw6dPhy4vBp2vRW6uzw.png" alt="Pandas" width="500"/>


Pandas are adorable little animals. But, they are also an open-source **data analysis and manilpulation tool**. It's built on top Python. Programmers like animals — can you tell?


**You can think of Pandas as cuter, better and more powerful version EXCEL**

## Importing the Pandas Library

**Pandas** is a library that is built on top of the **Numpy** library.

**Libraries:** are a collection of functions on methods that allow us to form specific actions 


---



Before we can use Pandas and Numpy, we need to **import** their libraries in Python. 



This is the standard way of importanting the Pandas and Numpy Library

When we import pandas as **pd**, every time we want to access the pandas library we can just write "pd" instead of "pandas". Likewise, importing **numpy** as np, allows us to access the numpy library with the shorter **np**


In [ ]:
import pandas as pd    
import numpy as np

## Reading a Excel (.xlsx) and CSV (.csv) Files

We can use Pandas to read and import data in different file formats.

Two common data file formats are **(1) Excel** and **(2) CSV**. 


### Excel

<img src="https://drive.google.com/uc?id=1aB2K0YSZb4Zx-eeUpkS2cdtRH3AP5tOS" alt="Excel" width="1000"/>


Excel (.xlsx) files are workbook files that are specifically created for Microsoft Excel, but they work in Google Sheets as well.  

---

#### Reading Excel Files 

To import excel files in Pandas we use the [pd.read_excel](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html) function


> df = pd.read_excel ("path-of-file.xlsx")


In this case:



*   **df** is the variable that we are saving the data frame to
*   **=** is how we assign data to a variable
*   **pd** is the name of the *library*
*   **read_excel** is the name of the *function*
*   **path-of-file.xlsx** is the *parameter* (i.e. the name of the file)

### Exercise 1: Reading your first excel file

Try importing your first excel file. The name of the file is **https://github.com/AutismResearchCentre/Spark_2020_Day_01/blob/master/US_School_Census_2019.xlsx?raw=true** (*note* The name of the file is in quotations because it string. We put all strings in single quotes ('string') or double quotes ("string"). Save the string to a variable called **census_2019**.



In [ ]:
# TO-DO: Read in the US School Census Data file from 2019

path = "https://github.com/AutismResearchCentre/Spark_Datasets/blob/master/US_School_Census_2019.xlsx?raw=true"

# [VARIABLE_NAME] = pd.read_excel([PATH_NAME])
census_2019 = pd.read_excel("https://github.com/AutismResearchCentre/Spark_Datasets/blob/master/US_School_Census_2019.xlsx?raw=true")


To view the data frame just call the name of the variable that you saved the data frame to.

In [ ]:
# Type the name of the variable of the data frame and run the cell

census_2019

Country Region  ...   Role_Model_Type       Charity_Donation
0       USA     AK  ...          Relative      Wildlife, animals
1       USA     NJ  ...             Other      Wildlife, animals
2       USA     OH  ...          Relative                 Health
3       USA     PA  ...               NaN                    NaN
4       USA     NC  ...            Friend                 Health
..      ...    ...  ...               ...                    ...
245     USA     OH  ...     Sports person                 Health
246     USA     AZ  ...            Friend            Environment
247     USA     MO  ...               NaN                    NaN
248     USA     OH  ...  Religious figure            Environment
249     USA     CA  ...     Sports person  Arts, culture, sports

[250 rows x 60 columns]

### CSV


<img src="https://drive.google.com/uc?id=18Z5jhfYl-po6XXJUqgsA0bqfqVAiSma7" alt="CSV" width="1000"/>

CSV stands for Comma Separated Values. It is a plain text file where the values are separated by commas. CSV files can be opened by any spreadsheet program (e.g., Excel, Open Office, Google Sheets). You can also open and edit CSV files in simple text editors as well (e.g., NotePad)


#### Reading Excel Files 

To import excel files in Pandas we use the [pd.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function


`df = pd.read_csv ("path-of-file.csv")`


In this case:

*   **df** is the variable that we are saving the data frame to
*   **=** is how we assign data to a variable
*   **pd** is the name of the *library*
*   **read_csv** is the name of the *function*
*   **path-of-file.csv** is the *parameter* (i.e. the name of the file)


### Exercise 2: Reading your first csv file

Try importing your first csv file. It's very similar to reading an excel file. The path of the file is **https://raw.githubusercontent.com/AutismResearchCentre/Spark_Datasets/master/US_School_Census_2020.csv** Save the string to a variable called **census_2020**.

BONUS: What do you think would happen if we tried to use pd.read_csv to read an excel file?

In [ ]:
# TO-DO: Read your first CSV file below

path2 = "https://raw.githubusercontent.com/AutismResearchCentre/Spark_Datasets/master/US_School_Census_2020.csv"
census_2020 = pd.read_csv(path2)


## Viewing Data Frames

The DataFrame is a powerful tool in Pandas. To be able to use it properly, we need to know how to view the data in our DataFrames.

As seen above, simply calling the name of our DataFrame we can see our rows and columns. If our DataFrame is super large, Jupyter NoteBook will just display the first and last rows and columns.


In [ ]:
census_2020 = pd.read_csv("https://raw.githubusercontent.com/AutismResearchCentre/Spark_Datasets/master/US_School_Census_2020.csv")


In [ ]:
# Simply call the name of the DataFrame to display table

census_2020

Country Region  ...   Role_Model_Type       Charity_Donation
0       USA     AK  ...          Relative      Wildlife, animals
1       USA     NJ  ...             Other      Wildlife, animals
2       USA     OH  ...          Relative                 Health
3       USA     PA  ...               NaN                    NaN
4       USA     NC  ...            Friend                 Health
..      ...    ...  ...               ...                    ...
245     USA     OH  ...     Sports person                 Health
246     USA     AZ  ...            Friend            Environment
247     USA     MO  ...               NaN                    NaN
248     USA     OH  ...  Religious figure            Environment
249     USA     CA  ...     Sports person  Arts, culture, sports

[250 rows x 60 columns]

### Shape of the Data Frame

How do we know how many data points (rows) we have? How do we know how many features (columns) we have?

To determine how much data we have, we can use [df.shape](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html), which will give us information of how 



```
df.shape
```

*   **df** is the name of the Pandas DataFrame
*   **.shape** is the attribute of the data frame


df.shape will return a tuple in the form of (x, y)

*   **x** represents the number of rows
*   **y** represents the number of columns






In [ ]:
census_2020.shape

(250, 60)

Therefore, our data frame has **250 rows** and **60 columns**

### Head

To view the first rows of a DataFrame, we can use [pd.head()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html)



---


```
df.head()
```

By deafult` pd.head()` will display thie** first 5 rows** of the DataFrame




```
df.head(n)
```

Where **n is an integer** value that represents the number of rows. For example, `pd.head(15)` will display the first 15 rows. 





In [ ]:
census_2020.head()

Country Region  DataYear  ...  Preferred_Status Role_Model_Type   Charity_Donation
0     USA     AK      2020  ...              Rich        Relative  Wildlife, animals
1     USA     NJ      2020  ...              Rich           Other  Wildlife, animals
2     USA     OH      2020  ...             Happy        Relative             Health
3     USA     PA      2020  ...               NaN             NaN                NaN
4     USA     NC      2020  ...             Happy          Friend             Health

[5 rows x 60 columns]

### Tail

To view the last rows of a DataFrame, we can use [pd.tail()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.tail.html)

```
df.tail()
```

By deafult` df.tail()` will display by default **last 5 rows** of the DataFrame




```
df.tail(n)
```

Where **n is an integer** value that represents the number of rows. For example, `df.tail(15)` will display the last 15 rows. 





### Exercise 3: Males vs Females

Display the last 6 rows of the DataFrame? How many of the students are female? How many are male?

In [ ]:
# TO-DO: Display the last 6 rows of the Census_2020 DataFrame

census_2020.tail(6)



Country Region  ...   Role_Model_Type       Charity_Donation
244     USA     CA  ...          Relative      Wildlife, animals
245     USA     OH  ...     Sports person                 Health
246     USA     AZ  ...            Friend            Environment
247     USA     MO  ...               NaN                    NaN
248     USA     OH  ...  Religious figure            Environment
249     USA     CA  ...     Sports person  Arts, culture, sports

[6 rows x 60 columns]

In [ ]:
# TO-DO: How many students are female

print ('3 students are female, and 3 students are male') # Print statments let us see string in the console

3 students are female, and 3 students are male


# Indexing DataFrames





## DataFrame Index

The **DataFrame index** is the row labels of the table. 


*   DataFrame Indexes could be **numbers, words, or phrases**
*   DataFrame Indexes have to be **unique** (you can think of it as an ID number). We can't have two people with the same ID or name because that would be confusing.







### Integer DataFrame Index
For the Census 2020. Data we can see that that the DataFrame index is the bolded values **(0, 1, 2, 3 ...)**. This helps us identify each row.

In [ ]:
census_2020.head()

Country Region  DataYear  ...  Preferred_Status Role_Model_Type   Charity_Donation
0     USA     AK      2020  ...              Rich        Relative  Wildlife, animals
1     USA     NJ      2020  ...              Rich           Other  Wildlife, animals
2     USA     OH      2020  ...             Happy        Relative             Health
3     USA     PA      2020  ...               NaN             NaN                NaN
4     USA     NC      2020  ...             Happy          Friend             Health

[5 rows x 60 columns]

### String DataFrame Index

**String** is just another word another way to say a series of characters. 

Examples of strings include:
*   "Hello"
*   "Password123"
*   "12345"
*   'Bob the Builder!'
*   "$%&#$^#$^!*  *&%*salkdfjlksf"

Notice how strings are wrapped in double ("") or single quotations (''). 

Strings can also be DataFrame indexes. Take the example below of the cities DataFrame. The index of is the name of Finish cities (e.g., Helsink, Espoo, Tampere)


In [21]:
# Getting the cities DataFrame
cities = pd.DataFrame(['Helsinki', 'Espoo', 'Tampere', 'Vantaa', 'Oulu'])
cities['Population'] = [643272, 279044, 231853, 223027, 201810]
cities['Total area'] = [715.48, 528.03, 689.59, 240.35, 3817.52]
cities.set_index([0], inplace=True)

# Viewing the cities DataFrame
cities

Population  Total area
0                               
Helsinki      643272      715.48
Espoo         279044      528.03
Tampere       231853      689.59
Vantaa        223027      240.35
Oulu          201810     3817.52

### Setting a DataFrame Index

We can actually change the index of DataFrames. [df.set_index](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.set_index.html) allows us to set the DataFrame index (row labels) with existing columns

```
df.set_index('[COLUMN_NAME]', inplace=True)
```
*   **df** is the name of the DataFrame
*   **.set_index** is the function that we are apply to the dataframe
*   **'[COLUMN_NAME]'** is the name of the pre-existing column that we would 
*   **inplace** gives us the option to save over our pre-exisiting DataBase. If we set *inplace=True*, our edited DataFrame overwrites the original. If we set *inplace=False*, then we keep our original DataFrame. By deafult, inplace is set to False.















In [ ]:
# Changing the index from city name to population

cities.set_index('Population')

Total area
Population            
643272          715.48
279044          528.03
231853          689.59
223027          240.35
201810         3817.52

In [ ]:
# The original cities still exists

cities

Population  Total area
0                               
Helsinki      643272      715.48
Espoo         279044      528.03
Tampere       231853      689.59
Vantaa        223027      240.35
Oulu          201810     3817.52

In [ ]:
# Now lets use set_index with the inplace attribute
cities.set_index('Population', inplace=True)



In [ ]:
# When we call cities we can see that we have saved our changes

cities

Total area
Population            
643272          715.48
279044          528.03
231853          689.59
223027          240.35
201810         3817.52

### Exercise 3: Mountains! 

Below is a table about some mountains. What column be an appropriate index for the data? Why? Create a new index of the mtns DataFrame. (Do not forget to use the `inplace=True` attribute). 

In [22]:
# TO-DO: Run this cell to get assign the mtns DataFrame
mtns = pd.DataFrame([
    {'name': 'Mount Everest',
        'height (m)': 8848,
        'summited': 1953,
        'mountain range': 'Mahalangur Himalaya'},
    {'name': 'K2',
        'height (m)': 8611,
        'summited': 1954,
        'mountain range': 'Baltoro Karakoram'},
    {'name': 'Kangchenjunga',
        'height (m)': 8586,
        'summited': 1955,
        'mountain range': 'Kangchenjunga Himalaya'},
    {'name': 'Lhotse',
        'height (m)': 8516,
        'summited': 1956,
        'mountain range': 'Mahalangur Himalaya'},
])

mtns

name  height (m)  summited          mountain range
0  Mount Everest        8848      1953     Mahalangur Himalaya
1             K2        8611      1954       Baltoro Karakoram
2  Kangchenjunga        8586      1955  Kangchenjunga Himalaya
3         Lhotse        8516      1956     Mahalangur Himalaya

In [24]:
# TO-DO: Set an new index for the table

mtns = pd.DataFrame([
    {'name': 'Mount Everest',
        'height (m)': 8848,
        'summited': 1953,
        'mountain range': 'Mahalangur Himalaya'},
    {'name': 'K2',
        'height (m)': 8611,
        'summited': 1954,
        'mountain range': 'Baltoro Karakoram'},
    {'name': 'Kangchenjunga',
        'height (m)': 8586,
        'summited': 1955,
        'mountain range': 'Kangchenjunga Himalaya'},
    {'name': 'Lhotse',
        'height (m)': 8516,
        'summited': 1956,
        'mountain range': 'Mahalangur Himalaya'},
])

mtns.set_index('name',inplace=True)
mtns

height (m)  summited          mountain range
name                                                       
Mount Everest        8848      1953     Mahalangur Himalaya
K2                   8611      1954       Baltoro Karakoram
Kangchenjunga        8586      1955  Kangchenjunga Himalaya
Lhotse               8516      1956     Mahalangur Himalaya

## Loc and Label-based Indexing

[df.loc[]](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) is used to slice and index a DataFrame based on the labels of the columns and the rows.

The inputs of .loc can be:


---




### Using Loc for a Single Row

```
df.loc['ROW_LABEL']
```
Returns a single row of the DataFrame




In [26]:
# Read the csv file
hospitals = pd.read_csv("https://raw.githubusercontent.com/AutismResearchCentre/Spark_Datasets/master/Toronto_Hospitals.csv")

# Set the index to Hospital "Name"
hospitals.set_index('Name', inplace=True)

# View hospital DataFrame
hospitals

Founded  ...                                     Former name(s)
Name                                       ...                                                   
Baycrest                             1918  ...                      Toronto Jewish Old Folks Home
Bellwood                             1984  ...                                                NaN
Bridgepoint                          1875  ...                                 Riverdale Hospital
Casey House                          1988  ...                                                NaN
CAMH College                         1966  ...                     Clarke Institute of Psychiatry
CAMH Queen                           1850  ...  Provincial Lunatic Asylum;Queen Street Mental ...
CAMH Russell                         1949  ...                      Addiction Research Foundation
Centric Health Surgical Centre       1960  ...                            Don Mills Surgical Unit
Etobicoke General                    1972  ...                                                NaN
Holland Bloorview                    1899  ...  Home for Incurable Children;Bloorview MacMilla...
Hospital for Sick Children           1875  ...                Victoria Hospital for Sick Children
Humber River                         1997  ...                                                NaN
Michael Garron                       1929  ...                      Toronto East General Hospital
Mount Sinai                          1923  ...  Toronto Hebrew Maternity and Convalescent Hosp...
North York General                   1968  ...                                                NaN
Princess Margaret Cancer Centre      1952  ...  Ontario Cancer Institute;Princess Margaret Hos...
Providence Healthcare                1957  ...  House of Providence;Providence Villa and Hospi...
Queensway Health Centre              1956  ...                         Queensway General Hospital
Centenary                            1967  ...  Centenary Health Centre;Scarborough Centenary ...
Runnymede Healthcare Centre          1945  ...                                 Runnymede Hospital
Scarborough General                  1956  ...                       Scarborough General Hospital
Birchmount                           1985  ...                         Scarborough Grace Hospital
St. Joseph's Health Centre           1921  ...                              St. Joseph's Hospital
St. Michael's                        1892  ...                                                NaN
Sunnybrook                           1948  ...  Sunnybrook Military Hospital;Sunnybrook Hospit...
St. John's Rehab                     1937  ...                   St. John's Convalescent Hospital
Toronto General                      1812  ...                              York General Hospital
Toronto Grace Health Centre          1905  ...                             Toronto Grace Hospital
Toronto Rehabilitation Institute     1998  ...                                                NaN
Toronto Western                      1895  ...                                                NaN
West Park Healthcare Centre          1904  ...  Toronto Free Hospital for Consumptive Poor;Que...
Women's College                      1883  ...                                                NaN

[32 rows x 5 columns]

In [27]:
# Get the row for Holland Bloorview

hospitals.loc['Holland Bloorview'] # Case Sensitve

Founded                                                              1899
District                                                       North York
Network                                                               NaN
University affiliate              University of Toronto(fully-affiliated)
Former name(s)          Home for Incurable Children;Bloorview MacMilla...
Name: Holland Bloorview, dtype: object

In [28]:
# Get the row for Centric Health Surgical Centre

hospitals.loc['Centric Health Surgical Centre']

Founded                                    1960
District                             North York
Network                                     NaN
University affiliate                        NaN
Former name(s)          Don Mills Surgical Unit
Name: Centric Health Surgical Centre, dtype: object

### Using Loc for a Single Cell

```
df.loc[ROW_NAME, COLUMN_NAME]
```

Returns a specific cell of the DataFrame


In [ ]:
# Get the year Etobicoke General was founded

hospitals.loc["Etobicoke General", "Founded"]

1972

In [ ]:
# Get the district of Casey House 

hospitals.loc["Casey House", "District"]

'Old Toronto'

In [ ]:
# Get the former name of Bellwood 

hospitals.loc["Bellwood", "Former name(s)"] # Return nan, which indicates missing data


nan

### Using Loc to Subset the DataFrame

Sometimes we only want a subset of our DataFrame. 

#### The slice operator (:)

```
[START_INDEX : STOP_INDEX]
```

The slice operator cuts a sequence that starts at the `START_INDEX` and ends with `END_INDEX`


In [29]:
# Slicing Rows: Get rows Bellwood to Etobicoke General

hospitals.loc['Bellwood': 'Etobicoke General']

Founded  ...                                     Former name(s)
Name                                     ...                                                   
Bellwood                           1984  ...                                                NaN
Bridgepoint                        1875  ...                                 Riverdale Hospital
Casey House                        1988  ...                                                NaN
CAMH College                       1966  ...                     Clarke Institute of Psychiatry
CAMH Queen                         1850  ...  Provincial Lunatic Asylum;Queen Street Mental ...
CAMH Russell                       1949  ...                      Addiction Research Foundation
Centric Health Surgical Centre     1960  ...                            Don Mills Surgical Unit
Etobicoke General                  1972  ...                                                NaN

[8 rows x 5 columns]

In [30]:
# Slicing Rows and Columns: 

# Get Rows from Casey House to Holland Bloorview, and columns from District to Network

hospitals.loc['Casey House': 'Holland Bloorview', 'District' : 'Network']

District                                 Network
Name                                                                               
Casey House                     Old Toronto                                     NaN
CAMH College                    Old Toronto  Centre for Addiction and Mental Health
CAMH Queen                      Old Toronto  Centre for Addiction and Mental Health
CAMH Russell                    Old Toronto  Centre for Addiction and Mental Health
Centric Health Surgical Centre   North York                                     NaN
Etobicoke General                 Etobicoke             William Osler Health System
Holland Bloorview                North York                                     NaN


```
[:]
```

A single colon (with nothing before or after it), means get everything

In [ ]:
# Get the whole hospital DataFrame

hospitals.loc[:] # This is the same as just calling hospitals

Founded  ...                                     Former name(s)
Name                                       ...                                                   
Baycrest                             1918  ...                      Toronto Jewish Old Folks Home
Bellwood                             1984  ...                                                NaN
Bridgepoint                          1875  ...                                 Riverdale Hospital
Casey House                          1988  ...                                                NaN
CAMH College                         1966  ...                     Clarke Institute of Psychiatry
CAMH Queen                           1850  ...  Provincial Lunatic Asylum;Queen Street Mental ...
CAMH Russell                         1949  ...                      Addiction Research Foundation
Centric Health Surgical Centre       1960  ...                            Don Mills Surgical Unit
Etobicoke General                    1972  ...                                                NaN
Holland Bloorview                    1899  ...  Home for Incurable Children;Bloorview MacMilla...
Hospital for Sick Children           1875  ...                Victoria Hospital for Sick Children
Humber River                         1997  ...                                                NaN
Michael Garron                       1929  ...                      Toronto East General Hospital
Mount Sinai                          1923  ...  Toronto Hebrew Maternity and Convalescent Hosp...
North York General                   1968  ...                                                NaN
Princess Margaret Cancer Centre      1952  ...  Ontario Cancer Institute;Princess Margaret Hos...
Providence Healthcare                1957  ...  House of Providence;Providence Villa and Hospi...
Queensway Health Centre              1956  ...                         Queensway General Hospital
Centenary                            1967  ...  Centenary Health Centre;Scarborough Centenary ...
Runnymede Healthcare Centre          1945  ...                                 Runnymede Hospital
Scarborough General                  1956  ...                       Scarborough General Hospital
Birchmount                           1985  ...                         Scarborough Grace Hospital
St. Joseph's Health Centre           1921  ...                              St. Joseph's Hospital
St. Michael's                        1892  ...                                                NaN
Sunnybrook                           1948  ...  Sunnybrook Military Hospital;Sunnybrook Hospit...
St. John's Rehab                     1937  ...                   St. John's Convalescent Hospital
Toronto General                      1812  ...                              York General Hospital
Toronto Grace Health Centre          1905  ...                             Toronto Grace Hospital
Toronto Rehabilitation Institute     1998  ...                                                NaN
Toronto Western                      1895  ...                                                NaN
West Park Healthcare Centre          1904  ...  Toronto Free Hospital for Consumptive Poor;Que...
Women's College                      1883  ...                                                NaN

[32 rows x 5 columns]

In [ ]:
# Get all the rows, but only the 'Network' and 'Former name(s)' Columns

hospitals.loc[:, 'Network': 'Former name(s)']

Network  ...                                     Former name(s)
Name                                                                      ...                                                   
Baycrest                                                             NaN  ...                      Toronto Jewish Old Folks Home
Bellwood                                         Edgewood Health Network  ...                                                NaN
Bridgepoint                                          Sinai Health System  ...                                 Riverdale Hospital
Casey House                                                          NaN  ...                                                NaN
CAMH College                      Centre for Addiction and Mental Health  ...                     Clarke Institute of Psychiatry
CAMH Queen                        Centre for Addiction and Mental Health  ...  Provincial Lunatic Asylum;Queen Street Mental ...
CAMH Russell                      Centre for Addiction and Mental Health  ...                      Addiction Research Foundation
Centric Health Surgical Centre                                       NaN  ...                            Don Mills Surgical Unit
Etobicoke General                            William Osler Health System  ...                                                NaN
Holland Bloorview                                                    NaN  ...  Home for Incurable Children;Bloorview MacMilla...
Hospital for Sick Children                                           NaN  ...                Victoria Hospital for Sick Children
Humber River                                       Humber River Hospital  ...                                                NaN
Michael Garron                               Toronto East Health Network  ...                      Toronto East General Hospital
Mount Sinai                                          Sinai Health System  ...  Toronto Hebrew Maternity and Convalescent Hosp...
North York General                                                   NaN  ...                                                NaN
Princess Margaret Cancer Centre                University Health Network  ...  Ontario Cancer Institute;Princess Margaret Hos...
Providence Healthcare                               Unity Health Toronto  ...  House of Providence;Providence Villa and Hospi...
Queensway Health Centre                         Trillium Health Partners  ...                         Queensway General Hospital
Centenary                                     Scarborough Health Network  ...  Centenary Health Centre;Scarborough Centenary ...
Runnymede Healthcare Centre                                          NaN  ...                                 Runnymede Hospital
Scarborough General                           Scarborough Health Network  ...                       Scarborough General Hospital
Birchmount                                    Scarborough Health Network  ...                         Scarborough Grace Hospital
St. Joseph's Health Centre                          Unity Health Toronto  ...                              St. Joseph's Hospital
St. Michael's                                       Unity Health Toronto  ...                                                NaN
Sunnybrook                                                           NaN  ...  Sunnybrook Military Hospital;Sunnybrook Hospit...
St. John's Rehab                                                     NaN  ...                   St. John's Convalescent Hospital
Toronto General                                University Health Network  ...                              York General Hospital
Toronto Grace Health Centre                           The Salvation Army  ...                             Toronto Grace Hospital
Toronto Rehabilitation Institute               University Health Network  ...                                                NaN
Toronto Western                                University Health N

```
[: END_INDEX]
```

Assumes the Start Index is the first entry of the DataFrame

```
[START_INDEX : ]
```

Assumes the End Index is  the last entry of the DataFrame


In [ ]:
# Get all rows from "Sunnybrook" to the end of the list

hospitals.loc['Sunnybrook':]

Founded  ...                                     Former name(s)
Name                                       ...                                                   
Sunnybrook                           1948  ...  Sunnybrook Military Hospital;Sunnybrook Hospit...
St. John's Rehab                     1937  ...                   St. John's Convalescent Hospital
Toronto General                      1812  ...                              York General Hospital
Toronto Grace Health Centre          1905  ...                             Toronto Grace Hospital
Toronto Rehabilitation Institute     1998  ...                                                NaN
Toronto Western                      1895  ...                                                NaN
West Park Healthcare Centre          1904  ...  Toronto Free Hospital for Consumptive Poor;Que...
Women's College                      1883  ...                                                NaN

[8 rows x 5 columns]

In [ ]:
# Get all rows up to "Bridgepoint" (Inclusive)

hospitals.loc[ :"Bridgepoint"] 

Founded  ...                 Former name(s)
Name                  ...                               
Baycrest        1918  ...  Toronto Jewish Old Folks Home
Bellwood        1984  ...                            NaN
Bridgepoint     1875  ...             Riverdale Hospital

[3 rows x 5 columns]

### Arrays for Indexing

Arrays can help us select the specifc items (columns or rows) that we need. 


```
[ INDEX1, INDEX2, INDEX3 ... INDEX8 ]
```

This helps us select the specific indexes that we want. 


```
[ INDEX13, INDEX2, INDEX34 ... INDEX2 ]
```

Indexes do not need to be in order

In [ ]:
# Get the Years that Sunnybrook, Humber River, Mount Sinai were founded 

hospitals.loc[['Sunnybrook', 'Humber River', 'Mount Sinai'], 'Founded'] # Notice how he names of the hospitals are in an array

Name
Sunnybrook      1948
Humber River    1997
Mount Sinai     1923
Name: Founded, dtype: int64

In [ ]:
# We can also save the array of of indexes to variables

row_indexes = ['Birchmount', 'Baycrest', 'Toronto General']
column_indexes = ['District', 'Founded']

hospitals.loc[row_indexes, column_indexes] 


District  Founded
Name                                 
Birchmount       Scarborough     1985
Baycrest          North York     1918
Toronto General  Old Toronto     1812

### Exercise 4: Pokedex (Indexing with Pokemon)

Below is a table of Generation 1 Pokemon. Answer the question below. 

In [51]:
pokemon =  pd.read_csv("https://raw.githubusercontent.com/AutismResearchCentre/Spark_2020_Day_01/master/Gen1_Pokemon.csv")
pokemon

Name  Number     Type       Evolves_Into
0     Bulbasaur       1    Grass     Ivysaur (#002)
1       Ivysaur       2    Grass    Venusaur (#003)
2      Venusaur       3    Grass     Mega Evolution
3    Charmander       4     Fire  Charmeleon (#005)
4    Charmeleon       5     Fire   Charizard (#006)
..          ...     ...      ...                ...
146     Dratini     147   Dragon   Dragonair (#148)
147   Dragonair     148   Dragon   Dragonite (#149)
148   Dragonite     149   Dragon   End of evolution
149      Mewtwo     150  Psychic     Mega Evolution
150         Mew     151  Psychic   Unable to evolve

[151 rows x 4 columns]

**Set the index of Pokemon DataFrame to `"Name"`**

In [52]:
# TO-DO: Set the `"Name"` as the DataFrame Index (Row Labels)

pokemon.set_index('Name',inplace=True)
pokemon

Number     Type       Evolves_Into
Name                                          
Bulbasaur        1    Grass     Ivysaur (#002)
Ivysaur          2    Grass    Venusaur (#003)
Venusaur         3    Grass     Mega Evolution
Charmander       4     Fire  Charmeleon (#005)
Charmeleon       5     Fire   Charizard (#006)
...            ...      ...                ...
Dratini        147   Dragon   Dragonair (#148)
Dragonair      148   Dragon   Dragonite (#149)
Dragonite      149   Dragon   End of evolution
Mewtwo         150  Psychic     Mega Evolution
Mew            151  Psychic   Unable to evolve

[151 rows x 3 columns]

**What is the type of Pokemon is the `"Abra"`?**

<img src="https://upload.wikimedia.org/wikipedia/en/e/ea/Pok%C3%A9mon_Abra%2C_Kadabra%2C_Alakazam_art.png" alt="Abra" width="300"/>

In [53]:
# TO-DO: Find the type of Abra?

pokemon.loc['Abra','Type']



'Psychic'

**What are the evolutions and numbers of `"Charmander"`, `"Pikachu"` and `"Squirtle"`?**

<img src="https://cdn.vox-cdn.com/thumbor/3ajecDMOIH59cbOeyO0bap_4wj4=/0x0:2257x1320/1200x800/filters:focal(949x480:1309x840)/cdn.vox-cdn.com/uploads/chorus_image/image/63738986/pokemon.0.0.png" alt="Abra" width="400"/>



In [59]:
# TO-DO: Get the evolutions of these three pokemon

pokemon.loc[['Charmander','Pikachu','Squirtle'],['Evolves_Into','Number']]




Evolves_Into  Number
Name                                 
Charmander  Charmeleon (#005)       4
Pikachu         Raichu (#026)      25
Squirtle     Wartortle (#008)       7



**Cut the Table from `"Eevee"` to `"Mewtwo"`. What is the range of numbers in the Pokedex**

<img src="https://upload.wikimedia.org/wikipedia/en/thumb/a/a9/Pok%C3%A9mon_Eevee_art.png/220px-Pok%C3%A9mon_Eevee_art.png" alt="Evee" width="100"/>

to 

<img src="https://upload.wikimedia.org/wikipedia/en/4/43/Pok%C3%A9mon_Mewtwo_art.png" alt="Evee" width="100"/>

In [61]:
# TO-DO: Cut the table from Evee to Dragonite

pokemon.loc['Eevee':'Dragonite']


Number      Type      Evolves_Into
Name                                          
Eevee          133    Normal    Multiple[nb 7]
Vaporeon       134     Water  End of Evolution
Jolteon        135  Electric  End of evolution
Flareon        136      Fire  End of evolution
Porygon        137    Normal   Porygon2 (#233)
Omanyte        138      Rock    Omastar (#139)
Omastar        139      Rock  End of evolution
Kabuto         140      Rock   Kabutops (#141)
Kabutops       141      Rock  End of evolution
Aerodactyl     142      Rock    Mega Evolution
Snorlax        143    Normal  End of evolution
Articuno       144       Ice  Unable to evolve
Zapdos         145  Electric  Unable to evolve
Moltres        146      Fire  Unable to evolve
Dratini        147    Dragon  Dragonair (#148)
Dragonair      148    Dragon  Dragonite (#149)
Dragonite      149    Dragon  End of evolution

# Cleaning Data

Before we can start analyzing data, we need to clean it. 

IBM Data Analytics estimates that data scientists can spend up to 80% of time cleaning data.

## Missing Values
One of the biggest data cleaning tasks is the problem of **missing values**.

### Sources of Missing Values

Data can be missing for a variety of different reasons. For example,

* The participant forget to fill in a field in a survey
* Data was lost while transferring files
* There was a programming error
* Data was not collected for a specific participant




We are going to load a csv file with some propery data. As you can see there is some issue with this Data.

![Imgur](https://i.imgur.com/cpDUkBk.png)

In [25]:
# Import libraries
import pandas as pd
import numpy as np

# Read csv file into a pandas DataFrame
df_property = pd.read_csv("https://raw.githubusercontent.com/nguyenjenny/spark_shared_repo/main/datasets/Property_Data.csv")

# Print out the Data
df_property

PID  ST_NUM     ST_NAME  ... NUM_BATH SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5    --   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700   NaN
4          NaN   203.0    BERKELEY  ...        2  1600   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800   NaN
6  100007000.0     NaN  WASHINGTON  ...   HURLEY   950   NaN
7  100008000.0   213.0     TREMONT  ...        1   NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800   NaN

[9 rows x 8 columns]

## Features and Data Types

Before we begin cleaning our data we need to understand what features we have (i.e. what do the columns of the data represent) and what are the expected data types.


### What are the features?
- `PID`: Property ID number
- `ST_NUM`: Street number
- `ST_NAME`: Street name
- `OWN_OCCUPIED`: Is the residence owner occupied
- `NUM_BEDROOMS`: Number of bedrooms
- `NUM_BATH`: Number of bathrooms
- `SQ_FT`: Square footage of the property
- `Extra`: Extra column with no data in it

### What are the expected types?
- `PID`: float (decimal number) or int (whole number), something numeric
- `ST_NUM`: float (decimal number) or int (whole number), something numeric
- `ST_NAME`: string (a mix of letters or numbers)
- `OWN_OCCUPIED`: string (where Y = "yes" and N "no")
- `NUM_BEDROOMS`: float (decimal number) or int (whole number), something numeric
- `NUM_BATH`: float (decimal number) or int (whole number), something numeric
- `SQ_FT`: float (decimal number) or int (whole number), something numeric
- `Extra`: NaN (missing data)

## How Pandas Deals with Missing Values (`NaN`)

Missing data in pandas is often called `NaN` which stands for **Not a Number**.

Here we can see that row `4`, column `"PID"` is  nan.




In [26]:
# Index row 4, column "PID"
df_property.loc[4, "PID"]

nan

In [27]:
## How to check if data is considered missing

## How to check if Data is Considered Missing (`isnull`)

There is a method called `isnull` that we can use to determine if data is missing.

- Missing or null data will return `True`
- Non-missing or non-null data will return `False`

In [28]:
# Check is the data is null or not
df_property.isnull()

PID  ST_NUM  ST_NAME  OWN_OCCUPIED  NUM_BEDROOMS  NUM_BATH  SQ_FT  Extra
0  False   False    False         False         False     False  False   True
1  False   False    False         False         False     False  False   True
2  False    True    False         False          True     False  False   True
3  False   False    False         False         False      True  False   True
4   True   False    False         False         False     False  False   True
5  False   False    False         False          True     False  False   True
6  False    True    False          True         False     False  False   True
7  False   False    False         False         False     False   True   True
8  False   False    False         False         False     False  False   True

We can also check if nulls on a single column

In [29]:
# We can check a single column
print(df_property['ST_NUM'])
df_property['ST_NUM'].isnull()

0    104.0
1    197.0
2      NaN
3    201.0
4    203.0
5    207.0
6      NaN
7    213.0
8    215.0
Name: ST_NUM, dtype: float64


0    False
1    False
2     True
3    False
4    False
5    False
6     True
7    False
8    False
Name: ST_NUM, dtype: bool

In [30]:
# We can also count how many missing/null values we have using the `sum()` method
df_property['ST_NUM'].isnull().sum()

2

In [31]:
# We can use the sum method on the entire DataFrame too
df_property.isnull().sum()

PID             1
ST_NUM          2
ST_NAME         0
OWN_OCCUPIED    1
NUM_BEDROOMS    2
NUM_BATH        1
SQ_FT           1
Extra           9
dtype: int64

## Standard Missing Values

Pandas will recognize the following automatically as standard missing values

- empty cells
- "n/a"
- "NA"
- "na"

As we can see all the items in blue, have been importated as `NaN`
![Imgur](https://i.imgur.com/cpDUkBk.png)

In [32]:
# Return the table
df_property

PID  ST_NUM     ST_NAME  ... NUM_BATH SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5    --   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700   NaN
4          NaN   203.0    BERKELEY  ...        2  1600   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800   NaN
6  100007000.0     NaN  WASHINGTON  ...   HURLEY   950   NaN
7  100008000.0   213.0     TREMONT  ...        1   NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800   NaN

[9 rows x 8 columns]

In [ ]:
# Non standard missing val

## Non-standard Missing Values

There are also some missing values that are not automatically recognized as missing. For example `"--"` is being recognized as a string instead of a missing value. 

In this case, we tell pandas what our missing values are by using the parameter `na_values`, when we use `pd.read_csv`.

In [33]:
# Create a list of missing values and save as a variable
missing_values = ['n/a', 'na', 'NaN', 'NA', '--']

# Read csv as as pandas data frame, but use the paramter na_values
df_property = pd.read_csv("https://raw.githubusercontent.com/nguyenjenny/spark_shared_repo/main/datasets/Property_Data.csv", na_values=missing_values)

# Return df_property, now "--" has been replaced with NaN

df_property


PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700.0   NaN
4          NaN   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...   HURLEY   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

## Dealing with Errors in the Data

There are also a few errors in the data (yellow cells). 

![Imgur](https://i.imgur.com/cpDUkBk.png)

We can use `replace` to mass replace incorrect cells. Or, we can use fix errors on a case-by-case basis by overwriting them using indexing (`loc`).

### Replacing Errors (`replace`)

This method works well, when there's a value in a data set that makes no sense. In the data set we have, "HURLEY" is incorrect, so we can replace with nan (`np.nan`)

We use the `replace` method.

```
df.replace("[OLD_VALUE]", "[NEW_VALUE]")
```

In [34]:
# Replace "Hurley" with np.nan
df_property = df_property.replace("HURLEY", np.nan)

# Show data frame
df_property

PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700.0   NaN
4          NaN   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...      NaN   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

### Overwriting Errors with Indexing (`loc`)

To overwrite a data point or error. We simply index it using `loc` and the make it equal to our new value.

```
df.loc[ "ROW_NAME" , "COLUMN_NAME"] = [NEW_VALUE]


```

We want to overwrite errors using indexing when we have to deal with data on a case by case basis.  For example, it doesn't make sense to have a value of `12` for the column `"OWN_OCCUPIED"`. But we don't want to just replace all values of `12` with `np.nan`. For example, we could have a house with a street number of 12 and that's not an error. This is where indexing comes in handy.  


The row name is `3` and the row_columns is `"OWN_OCCUPIED"` and we want to replace this with `np.nan` 

In [35]:
# Overwrite row 3, column "OWN_OCCUPIED" with np.nan
df_property.loc[3, "OWN_OCCUPIED"] = np.nan

# Show the updated dataframe
df_property

PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700.0   NaN
4          NaN   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...      NaN   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

### Example 5: Replacing missing data using overwriting with indexing

Overwriting with indexing can also be used to replacing missing values with actual values. 



Lets replace this missing value in column "PID" with 100005000


Lets replace this missing value in column `"PID"` with `100005000`

In [36]:
### TO-DO: Repalce this missing ID in the "PID" column with 100005000

df_property.loc[4,'PID']=100005000
df_property




PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...      NaN   700.0   NaN
4  100005000.0   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...      NaN   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

## Replacing Missing Data (`fillna`)

The `fillna` method is used to replace missing data. 

```
df = df.fillna(VALUE_TO_FILL)

df["COLUMN_NAME"] = df["COLUMN_NAME"].fillna(VALUE_TO_FILL)
```

Let's say we want to assume if we don't have information about the number of bathrooms (`"NUM_BATH"`), we can assume there is only 1 bathroom. 


In this case, we want to fill value to be 1. We can apply `fillna`

In [37]:
# Fill all nan values
df_property["NUM_BATH"] = df_property["NUM_BATH"].fillna(1)

# Show the new dataframe
df_property

PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...        1   700.0   NaN
4  100005000.0   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...        1   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

For number of bedrooms (`"NUM_BEDROOMS"`), we can fill missing data with the mean number of bedrooms

In [38]:
# Calculate the mean number of bedrooms first and save as variable 

mean_beds = df_property["NUM_BEDROOMS"].mean() # You will learn more about this later

print(mean_beds)

2.1666666666666665


In [39]:
# Fill all nan values for NUM_BEDRROMS with the mean

# Fill all nan values
df_property["NUM_BEDROOMS"] = df_property["NUM_BEDROOMS"].fillna(mean_beds)

# Show the new dataframe
df_property



PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...        1   700.0   NaN
4  100005000.0   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...        1   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

### Example 6: Filling missing data for the `"OWN_OCCUPIED"` column

If data on whether a place is `"OWN_OCCUPIED"` is missing, fill it with `"N"`, meaning that it is not occupied.

In [40]:
### TO-DO: Fill missing "OWN_OCCUPIED" data with "N"

df_property["OWN_OCCUPIED"] =df_property["OWN_OCCUPIED"].fillna("N")
df_property



PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...        1   700.0   NaN
4  100005000.0   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...        1   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

## Dropping rows/indexes and columns with missing data (`dropna`)

Sometimes we may not want to use rows/indexes or columns if data is missing. This is where the  `dropna` function comes in handy.

Documentation for this function is found here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html?highlight=dropna

Parameters: 


-  `axis`: the axis that you want to drop data on 
  - can be "index" or "columns"
  - `axis="index"` drops rows with missing data
  - `axis="columns"` drops columns with missing data

- `how`: how you want to drop the data
  - can be "any" or "all"
  - `how="any"` If any NA values are present, drop that row or column
  - `how="all"` If all values are NA, drop that row or column.



### Examples 
Drop columns where all the values are missing
```
df.dropna(axis="columns", how="all") 
```

Drop columns where there are any missing values
```
df.dropna(axis="columns", how="any") 
```

Drop rows/index where all the values are missing
```
df.dropna(axis="index", how="all") 
```

Drop rows/index where there are any missing values
```
df.dropna(axis="index", how="any") 
```



In [41]:
# Drop columns where all values are missing
df_property.dropna(axis="columns", how="all") 

PID  ST_NUM     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS NUM_BATH   SQ_FT
0  100001000.0   104.0      PUTNAM            Y      3.000000        1  1000.0
1  100002000.0   197.0   LEXINGTON            N      3.000000      1.5     NaN
2  100003000.0     NaN   LEXINGTON            N      2.166667        1   850.0
3  100004000.0   201.0    BERKELEY            N      1.000000        1   700.0
4  100005000.0   203.0    BERKELEY            Y      3.000000        2  1600.0
5  100006000.0   207.0    BERKELEY            Y      2.166667        1   800.0
6  100007000.0     NaN  WASHINGTON            N      2.000000        1   950.0
7  100008000.0   213.0     TREMONT            Y      1.000000        1     NaN
8  100009000.0   215.0     TREMONT            Y      2.166667        2  1800.0

In [42]:
# Drop columns where any values are missing
df_property.dropna(axis="columns", how="any")

PID     ST_NAME OWN_OCCUPIED  NUM_BEDROOMS NUM_BATH
0  100001000.0      PUTNAM            Y      3.000000        1
1  100002000.0   LEXINGTON            N      3.000000      1.5
2  100003000.0   LEXINGTON            N      2.166667        1
3  100004000.0    BERKELEY            N      1.000000        1
4  100005000.0    BERKELEY            Y      3.000000        2
5  100006000.0    BERKELEY            Y      2.166667        1
6  100007000.0  WASHINGTON            N      2.000000        1
7  100008000.0     TREMONT            Y      1.000000        1
8  100009000.0     TREMONT            Y      2.166667        2

#### Exercise 7: Drop rows/indexes where all and any of the data is missing.

In [45]:
### TO-DO: drop rows/indexes where all values are missing

df_property = df_property.dropna(axis="rows",how="all")
df_property



PID  ST_NUM     ST_NAME  ... NUM_BATH   SQ_FT Extra
0  100001000.0   104.0      PUTNAM  ...        1  1000.0   NaN
1  100002000.0   197.0   LEXINGTON  ...      1.5     NaN   NaN
2  100003000.0     NaN   LEXINGTON  ...        1   850.0   NaN
3  100004000.0   201.0    BERKELEY  ...        1   700.0   NaN
4  100005000.0   203.0    BERKELEY  ...        2  1600.0   NaN
5  100006000.0   207.0    BERKELEY  ...        1   800.0   NaN
6  100007000.0     NaN  WASHINGTON  ...        1   950.0   NaN
7  100008000.0   213.0     TREMONT  ...        1     NaN   NaN
8  100009000.0   215.0     TREMONT  ...        2  1800.0   NaN

[9 rows x 8 columns]

In [47]:
### TO-DO: drop rows/indexes where any values are missing
df_property = df_property.dropna(axis="rows",how="any")
df_property



Empty DataFrame
Columns: [PID, ST_NUM, ST_NAME, OWN_OCCUPIED, NUM_BEDROOMS, NUM_BATH, SQ_FT, Extra]
Index: []

# GitHub

<img src="https://www.c-sharpcorner.com/article/git-and-github-version-control-local-and-remote-repository/Images/Git%20And%20Github%20Version%20Control.png" alt="Colab" width="500"/>

**Git** is a version control programming language. 







## Installing Git and GitHub
    
 * **Installing Git on Windows**
      Download and install git from [Git for Windows](https://gitforwindows.org/)
 *  **Installing Git on Mac**
      Install [Git OSX Installer](https://sourceforge.net/projects/git-osx-installer/files/ ) 


2. Create a [GitHub Account](https://github.com/join) 
       

3. Configure Git to your GitHub account

  * Open git bash
  * Set a Git username 
    * `$ git config --global user.name "Mona Lisa"`
    * Confirm that you have set the Git username correctly
      * `$ git config --global user.name` 
      * `> Mona Lisa` 
  * Setting your commit email address in Git
    * `$ git config --global user.email "email@example.com"`
    * Confirm that you have set the Git email correctly
      * `$ git config --global user.email` 
      * `> email@example.com` 

4. Install **Git for Desktop**
* Install [git for desktop here](https://desktop.github.com/)


4. Share your git username with us in the Zoom chat and we will add you to repository

## Git Workflow

Below is an image of the git work flow and the common commands we use.

![](https://i.imgur.com/0T7u7bK.png)

### Remote Repository

A repo/repository is a place where all your files for a project are stored.

The remote repository is the repository shared with everyone and hosted on github.com. Because it's the shared version, we don't want to make changes to it directly.

### Local Repopository

The local repository is our version of the repository that we have saved on our personal comptuers. We can try things out and change things without affecting the remote version of the code that everyone else uses.  Once


### Working Directory

When we make changes to any of the the file in the repository. It goes into the the working directory. We can choose to keep (commit) or discard these changes

###  Staging Area

The staging area is where you want you add files that you have changed to get ready to commit

#### `git clone` 
To get a local repository we first need to clone it using the command git clone

#### `git add`
To add files to the staging area we use git add. This moves files from the working directory into the staging area. 

#### `git commit`
Once we have collected all the files that we would like to move you our local repository. We can use the command `git commit`. When we commit something, we must included a commit message that explains what the changes were. This helps us keep track of the changes incase we need to revert to an old version or solve a bug.

#### `git push`
Once we are comfortable with all our chages and commit and we are ready to share it with our team, we need `git push` our commits from the local repository to the remote repository

#### `git pull`
Other people will make changes to remote repository. To get those changes and to make sure our local repository is up-to-date, we need to `git pull` all the changes

#### `git checkout`
Git checkout is used in more complicated situations when we have to worry about branching

#### `git checkout`
Git checkout is used in more complicated situations when we have to worry about branching. It is used to change branches

#### `git merge`
Git merge is used when we want to combine separate branches into a single branch


## Cloning your first repository

1. Open **GitHub for Desktop**

2. Find the repository called **Shared Spark Repo** and select `Clone [your_username]/spark_shared_repo`

* ![image](https://i.imgur.com/7J9vRCl.png)

3. Clone the repository

* ![](https://i.imgur.com/2S3uPRy.png)

4.  Select `View the files of repository in Explorer` to see the files.

* You should be able to see all the cloned files saved locally
* ![](https://i.imgur.com/UlY2f8l.png)


We have just created our own local repository on our computer!

## Making your first commit

1. Open the README.md file located in the local repository in notebook or any text editor
2. Add your name under contributors and save the file

  * README files are in markdown. So `*` and a space represents a bullet point
  * ![](https://i.imgur.com/KtxIKam.png)

3. When you return to GitHub Desktop, you'll notice that the change will have been detected
  * Make sure the files you want to commit are ticked off. This move files from our working directory into the staging area.
  * Add a **commit message** at the bottom
    * Commit messages usually start with a **captialized word**
    * And start with a **verb in simple tense** (e.g., Add, Change, Remove, Fix)

  *![](https://i.imgur.com/zqx25UD.png)


We have just commit our changes to our local directory!



## Pushing the changes to remote repository

The changes we made are in our local repo are ready to be "pushed" to the main branch of your remote repo.

To push our changes to remote repository we can click any of the two circled push buttons.

![](https://i.imgur.com/RdX6Swn.png)

## Fetching and pulling up-to-date changes from the main remote repository

Before we make changes to our local repository, we want to get into the habit of fetching and pulling changes from the remote repository to make sure we are working with the most up-to-date version.

1. Fetch the from origin (i.e. the remote repository)
  * Fetch gets and list the commits that have been made on the remote branch
  * ![](https://i.imgur.com/BMckPsE.png)
2. To get those changes on our local machine, we need to pull them.
  * ![](https://i.imgur.com/4I9TQNk.png)

## Exercise 8: Adding your name to the contributor list

Go through all the steps listed above to add your name to contributor list. Since we are all editing the same README.md file, we want to make these changes one at a time.

**NB: Remember to `git pull` before you add your name, to get the most up-to-date version that your peers have added to**



## Exercise 9: Add your notebook from day 1 to GitHub Repository

There is a folder in the repository called `day01_notebooks`: https://github.com/nguyenjenny/spark_shared_repo/tree/main/day_01_notebooks

1. Download the notebook from Google Collab onto your computer as an .ipynb
   * ![](https://i.imgur.com/CHJDwJc.png)
   * Rename it `SPARK-Day1-[FirstName].ipynb`
2. Copy it to the correct `day01_notebooks` folder in your local repository
3. Stage and commit to your local repository
  * Be sure to add an informative message
4. Push the change to remote repository



## Accessing data files through GitHub

GitHub can be used to host and access datafiles through Google Collab.

Hotel booking demand data is hosted here: https://www.kaggle.com/jessemostipak/hotel-booking-demand?select=hotel_bookings.csv

We can download it to our local machine and copy it to our local repository, then commit and push the datasets to GitHub.

In the `datasets` folder of our shared repository (https://github.com/nguyenjenny/spark_shared_repo/tree/main/datasets), we have saved the hotel booking information in the file called Hotel_Booking.csv

* ![](https://i.imgur.com/5BmOZF7.png)


If we click `view raw`, it will lead us to the raw CSV file

* ![](https://i.imgur.com/vVEPQHK.png)

This leads us to the raw CSV file. We can use the URL as the file path when importing our data from pandas

* ![](https://i.imgur.com/VZfxDgv.png)

We can use `pd.read_csv("https://raw.githubusercontent.com/nguyenjenny/spark_shared_repo/main/datasets/Hotel_Bookings.csv")` to load our data


# Peer Programming: Loading and cleaning the hotel data 

## Part 1: Load the hotel booking data using `pd.read_csv` and save it under the varialbe `df_hotels`

The na_values are "non-standard". Look at the data to figure out what it is: https://raw.githubusercontent.com/nguyenjenny/spark_shared_repo/main/datasets/Hotel_Bookings.csv

*Hint there are multiple "non-standard" notations for missing data. It might help to look at the data in excel. *

In [ ]:
### TO-DO: Load the hotel booking data
import pandas as pd
import  numpy as np


### TO_DO: What are the missing values called in this data set
missing_values = [""] # replace with what the missing values are called (hint there are multiple nam)



### ### TO_DO: Import the data; use pd.read_csv to import the missing data, be sure to set `na_values = missing_values`
df_hotels = "" # replace with read statement





In [ ]:
missing_values = ["NULL", "missing_data"]
df_hotels = pd.read_csv("https://raw.githubusercontent.com/nguyenjenny/spark_shared_repo/main/datasets/Hotel_Bookings.csv", na_values = missing_values)

## Part 2: How many features (columns) and datapoints (rows/indexes) are three?



In [ ]:
# TO-DO: Get the number of features and datapoints





In [ ]:
df_hotels.shape

(119390, 32)

## Part 3: How much missing data do you have for each column?  Which column has the most missing data? Which column has the least missing data?

In [ ]:
### TO-DO: Figure out how much missing data there is in each column





In [ ]:
df_hotels.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

## Part 4: Use `loc` to get the `"hotel"`, `"adults"`, `"children"`, `"babies"` for the rows of `40600`, `40667`, `40679`  and `41160`. What do you notice about this data?



In [ ]:
### TO-DO: Use loc to get a subset of the dataset





In [ ]:
df_hotels.loc[[40600, 40667, 40679,  41160], ['hotel', 'adults', 'children', 'babies']]

hotel  adults  children  babies
40600  City Hotel       2       NaN       0
40667  City Hotel       2       NaN       0
40679  City Hotel       3       NaN       0
41160  City Hotel       2       NaN       0

## Part 5: Fill missing data for `"children"` with the value `0`. Confirm that there is no missing data using `isnull`

In [ ]:
### TO-DO: Fill missing data for the "children" column with the value 0






In [ ]:
### TO-DO: Confirm that there is now no missing data in that column






## Part 6: Drop all indexes/rows with any missing data save it as `df_hotels_processed`? How many data points do you have now?

---



In [ ]:
### TO-DO: Drop all indexes/rows with missing data

df_hotels_processed = "" # replace with drop statement




